### PR TITLE
x: add guppy lint to prevent move crates from depending on diem ones

### PR DESCRIPTION
**Pull request by vgao1996**
This adds a new lint preventing new Move-to-Diem dependencies from being added.

### CODE REVIEW

File: devtools/x/src/lint/guppy.rs [hunk 1]
```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
...
 }
+
+// Ensure that Move crates do not depend on Diem crates.
+#[derive(Debug)]
+pub struct MoveCratesDontDependOnDiemCrates;
+
+// List of existing move to diem dependencies.
+// The goal is eliminate all of them gradually.
+// In general, you should refrain from adding new entries to this list.
+const EXISTING_MOVE_TO_DIEM_DEPS: &[(&str, &str)] = &[
```
Comment (bmwill) [inline devtools/x/src/lint/guppy.rs:518]: Might be better to stick this in a field in x.toml?

File: devtools/x/src/lint/guppy.rs [hunk 2]
```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
...
+    ("compiled-stdlib", "diem-framework"),
+    ("compiled-stdlib", "diem-crypto"),
+    ("move-ir-types", "diem-types"),
+    ("move-lang", "fallible"),
+    ("move-lang", "diem-types"),
+    ("move-lang", "datatest-stable"),
+    ("move-lang-functional-tests", "functional-tests"),
+    ("move-lang-functional-tests", "diem-types"),
+    ("move-lang-functional-tests", "datatest-stable"),
+    ("move-lang-test-utils", "datatest-stable"),
```
Comment (bmwill) [inline devtools/x/src/lint/guppy.rs:543]: I think datatest-stable isn't really a "deim" dependency. Its only used by move tests IIRC

File: devtools/x/src/lint/guppy.rs [hunk 3]
```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
...
+    ("move-model", "diem-types"),
+    ("move-model", "datatest-stable"),
+    ("move-prover", "diem-types"),
+    ("move-prover", "diem-temppath"),
+    ("move-prover", "datatest-stable"),
+    ("abigen", "diem-types"),
+    ("abigen", "diem-temppath"),
```
Comment (bmwill) [inline devtools/x/src/lint/guppy.rs:550]: Whats the plan when we're using common deps like this? migrate to something on crates.io?

File: devtools/x/src/lint/guppy.rs [hunk 4]
```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
...
+    ("ir-to-bytecode-syntax", "diem-types"),
+    ("compiled-stdlib", "diem-types"),
+    ("compiled-stdlib", "diem-proptest-helpers"),
+    ("compiled-stdlib", "diem-framework"),
+    ("compiled-stdlib", "diem-crypto"),
+    ("move-ir-types", "diem-types"),
+    ("move-lang", "fallible"),
+    ("move-lang", "diem-types"),
+    ("move-lang", "datatest-stable"),
+    ("move-lang-functional-tests", "functional-tests"),
```
Comment (bmwill) [inline devtools/x/src/lint/guppy.rs:540]: Aren't "functional-tests" move functional tests?

File: devtools/x/src/lint/guppy.rs [hunk 5]
```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
...
+    }
+
+    false
+}
+
+fn is_existing_move_to_diem_dep(name1: &str, name2: &str) -> bool {
+    EXISTING_MOVE_TO_DIEM_DEPS
+        .iter()
+        .any(|(a, b)| *a == name1 && *b == name2)
+}
```
Comment (bmwill) [inline devtools/x/src/lint/guppy.rs:627]: This could be expensive, may the list could be changed to be a HashSet<(str, str)> instead?

**Review by bmwill — COMMENTED** [review]
Just some initial thoughts

File: devtools/x/src/lint/guppy.rs (hunk 1 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs:518]: Didn't know you could do that! Let me try

File: devtools/x/src/lint/guppy.rs (hunk 4 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs:540]: It's a bit complicated. Long story short:
1. The framework (`functional-tests`) itself is more of a Diem crate than a Move one.
2. The tests are mixed -- they contain both Move and Diem tests.

Eventually we'll need to split the framework into a Move one and a Diem one and migrate all the move tests. For now the crate should be marked as a Diem crate.

File: devtools/x/src/lint/guppy.rs (hunk 3 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs:550]: Yeah we'll likely migrate to the crates.io versions.

File: devtools/x/src/lint/guppy.rs (hunk 2 above)
Comment (metajack) [reply devtools/x/src/lint/guppy.rs:543]: diem-framework also uses it.

Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs:543]: So here what matters to us is not whether a crate is conceptually Diem but whether it resides in the Diem repo. I don't see us brining datatest-stable to the new repo so this is also a dependency we have to remove. 

(BTW what was the original reason we forked datatest in the first place? Did it have something to do with the TCB? cc @sunshowers)

Comment (sunshowers) [reply devtools/x/src/lint/guppy.rs:543]: We reimplemented datatest to make it work on stable rust.

Comment (sunshowers) [reply devtools/x/src/lint/guppy.rs:543]: datatest-stable can move into the diem-devtools repo

File: devtools/x/src/lint/guppy.rs (hunk 4 above)
Comment (metajack) [reply devtools/x/src/lint/guppy.rs:540]: These do appear only used by ir-testsutie and move-lang-functional-tests.

File: devtools/x/src/lint/guppy.rs (hunk 3 above)
Comment (metajack) [reply devtools/x/src/lint/guppy.rs:550]: Is diem-temppath the only example? As I recall that's a really minimal wrapper function for temppath, so maybe we should just get rid of it? :)

File: devtools/x/src/lint/guppy.rs (hunk 4 above)
Comment (metajack) [reply devtools/x/src/lint/guppy.rs:540]: If there are a subset of tests that should be diem-functional-tests, then I still think that functional-tests should be under Move, since it is used by ir-testsuite too. In other words, it's a functional test framework for move source/ir testing, that diem can import.

File: devtools/x/src/lint/guppy.rs (hunk 2 above)
Comment (metajack) [reply devtools/x/src/lint/guppy.rs:543]: In this case, since it probably does not need updating, we could just have it internal to both places. Or publish it seprately on crates.io and put it in its own repo.

File: devtools/x/src/lint/guppy.rs (hunk 4 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs:540]: > In other words, it's a functional test framework for move source/ir testing, that diem can import.
Unfortunately this is not the case. The existing functional test framework is highly coupled with Diem concepts like transactions, gas, state views etc. The Move framework we'll implement does not need and must not contain any of these -- it will mostly be a fresh rewrite with a more simplified model, borrowing some of the ideas. 

In other words, the existing functional test crate will stay in the Diem repo and should be treated as a Diem thing. With that being said I should probably label "move-lang-functional-tests" as Diem to avoid any confusion here.

File: devtools/x/src/lint/guppy.rs (hunk 3 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs:550]: Also `fallible`, `diem-fallible`, `num-variants` and `datatest-stable`? `diem-proptest-helpers` might also be tricky...

File: devtools/x/src/lint/guppy.rs (hunk 1 above)
Comment (sunshowers) [reply devtools/x/src/lint/guppy.rs:518]: Yeah it would be great to have this in x.toml :)

File: devtools/x/src/lint/guppy.rs [hunk 6]
```diff
@@ -507,3 +507,167 @@ impl PackageLinter for OnlyPublishToCratesIo {
...
+        let crate_name = metadata.name();
+        let crate_path = metadata.source().to_string();
+
+        if is_move_crate(&crate_path, crate_name) {
+            for direct_dep in metadata.direct_links() {
+                let dep = direct_dep.to();
+                let dep_name = dep.name();
+
+                if dep.in_workspace()
+                    && dep_name != "diem-workspace-hack" // Exclude `diem-workspace-hack` to reduce noise.
```
Comment (sunshowers) [inline devtools/x/src/lint/guppy.rs:652]: Hmm, you could avoid hardcoding it through https://github.com/diem/diem/blob/be4811ee515407175be79cbfcc333fa43ae548d9/devtools/x/src/lint/guppy.rs#L226-L233, but that's honestly much more complicated than it should be. I should fix this.

**Review by sunshowers — COMMENTED** [review]
Looks great modulo existing comments :)

File: devtools/x/src/lint/guppy.rs (hunk 3 above)
Comment (bmwill) [reply devtools/x/src/lint/guppy.rs:550]: aside from datatest i think you should be able to depend directly on existing published crates or just handle the logic yourself in the move crates

File: devtools/x/src/lint/guppy.rs (hunk 1 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs]: Done.

File: devtools/x/src/lint/guppy.rs (hunk 3 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs]: Yeah these dependencies shouldn't be a big problem to us. We have many choices.

File: devtools/x/src/lint/guppy.rs (hunk 6 above)
Comment (vgao1996) [reply devtools/x/src/lint/guppy.rs]: Moved it into x.toml.

**Comment by vgao1996** [comment]
@bmwill @metajack @sunshowers I've implemented the changes you suggested. Let me know if it looks good!

**Comment by sunshowers** [comment]
Looks good!

**Review by sunshowers — APPROVED** [review]
